articles/actions.md

```diff
@@ -0,0 +1,363 @@
+The result will indicate the final status of the goal and any user defined data as part of the [Action Interface Definition](#action-interface-definition).
+
+The server should cache the result once it is ready so multiple clients have to opportunity to get it.
+This is useful for debugging/introspection tools.
+To free up resources, the server should discard the result after a configurable timeout period.
+
+### Goal Status Topic
+
+* **Direction**: server publishes
+* **Content**: list of in-progress goals with goal ID, time accepted, and an enum indicating the status
+
+This topic is published by the server to broadcast the status of goals it has accepted.
+The purpose of the topic is for introspection; it is not used by the action client.
+Messages are published when transitions from one status to another occur.
+
+The QoS settings for an action server can be configured by the user.
+The default QoS settings for a DDS middleware should be TRANSIENT LOCAL with a history size of 1.
+This allows new subscribers to always get the latest state.
+
+The possible statuses are:
+
+- *Accepted* - The goal has been accepted by the action server
+- *Executing* - The goal is currently being executing by the action server
+- *Canceling* - The action server will try to cancel the indicated goal
+- *Succeeded* - The action server successfully reached the goal
+- *Aborted* - The action server failed reached the goal
+- *Canceled* - The action server successfully canceled the goal
+
+### Feedback Topic
+
+- **Direction**: server publishes
+- **Content**: goal ID, user defined feedback message
+
+This topic is published by the server to send progress feedback about the goal that is user-defined as part of the [Action Interface Definition](#action-interface-definition).
+It is up to the author of the action server to decide how often to publish the feedback.
+
+The QoS settings for feedback coming from an action server can be configured by the user.
+It is up to the clients to use compatible QoS settings.
+
+### Client/Server Interaction Examples
+
+Here are a couple of sequence diagrams depicting typical interactions between an action client and action server.
+
+#### Example 1
+
+In this example, the action client request a goal and gets a response from the server accepting the goal (synchronous).
+Upon accepting the goal, the action server starts a user defined execution method for completing the goal.
+Following the goal request, the client makes an asynchronous request for the result.
+The user defined method publishes feedback to the action client as it executes the goal.
+Ultimately, the user defined method populates a result message that is used as part of the result response.
+
+![Goal Lifecycle Example 0](../img/actions/interaction_example_0.png)
+
+#### Example 2
+
+This example is almost identical to the first, but this time the action  client requests for the goal to be canceled mid-execution.
```

Comment:
No message goes from `handle_cancel` to the user-defined execution method.

Reply:
The _handle_cancel_ and _handle_goal_ are meant to represent potential user-defined callbacks. The "user-defined execution method" represents the actual execution of the goal (that is triggered upon accepting a goal request). E.g. the [execute method from the proposed rclcpp example](https://github.com/ros2/examples/blob/01f24c9c64003be894584c53085c3910ec2268b7/rclcpp/minimal_action_server/not_composable.cpp#L41).

Reply:
Nevertheless, the diagram needs to have some indication that the user-defined code receives notification about the goal being cancelled for it to match the text above:

> Note that the user defined method is allowed to perform any shutdown operations after the cancel request before returning with the cancellation result.

This is what is happening in [the example](https://github.com/ros2/examples/blob/01f24c9c64003be894584c53085c3910ec2268b7/rclcpp/minimal_action_server/not_composable.cpp#L55) with the check for whether the goal has been cancelled or not.

Reply:
Added an arrow in 0baa626